lisp/mastodon-notifications.el

```diff
@@ -0,0 +1,168 @@
+;; along with mastodon.el.  If not, see <http://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; mastodon-notification.el provides notification functions.
+
+;; todo: - integrate media
+;;       - extend notification types
+;;       - replace `You' with `your_display_name'(@`your_user_name')
+;;       - generalize mastodon-tl--more or rebind it in notifications
+;;         example
+;;       - fix thread access
+;;       - Add keybinding of mastodon-notifications--get to `N'
+
+;; ;; One approach for generalizing the get-more functionality
+;; (defun mastodon-tl--get-more (more-function id-fun goto-fun
+;;                               processing-function)
+;;   "can work for both oldert toots and newest toots"
+;;   (interactive)
+;;   (let* ((tl (mastodon-tl--timeline-name))
+;;       (id (funcall id-fun))
+;;       (json (funcall more-function tl id)))
+;;     (when json
+;;       (with-current-buffer (current-buffer)
+;;      (let ((inhibit-read-only t))
+;;        (goto-char (funcall goto-fun))
+;;        (funcall processing-function json))))))
+;;
+;; ;; Equivelent to the mastodon-tl--more
+;; (mastodon-tl--get-more 'mastodon-tl--more-json 'mastodon-tl--oldest-id
+;;                        point-max mastodon-tl--timeline)
+;;
+;; ;; Equivelent to the mastodon-tl--update
+;; (mastodon-tl--get-more 'mastodon-tl--update-json 'mastodon-tl--newest-id
+;;                        point-min mastodon-tl--timeline)
+
+
+;;; Code:
+
+;;(require 'mastodon-media)
+(require 'mastodon-http)
+(require 'mastodon-tl)
+
+(defgroup mastodon-notifications nil
+  "Mastodon Notifications."
+  :prefix "mastodon-notifications-"
+  :group 'mastodon)
+
+(defvar mastodon-notifications--types
+  '("mention" "follow" "favourite" "reblog" )
+  "A list of notification types that have been implemented.")
+
+(defun mastodon-notifications--byline-concat (toot message)
+  "Add byline for TOOT with MESSAGE."
+      (concat
+       " "
+       (propertize message 'face 'highlight)
+       " "
+       "You!"))
+
+(defun mastodon-notifications--byline (toot message)
+  "Generate byline from TOOT based on MESSAGE for notification."
+  (let ((id (cdr (assoc 'id toot)))
+        (faved (mastodon-tl--field 'favourited toot))
+        (boosted (mastodon-tl--field 'reblogged toot)))
+    (propertize
+     (concat (propertize "\n | " 'face 'default)
+             (when boosted
+               (format "(%s) " (propertize "B" 'face 'success)))
+             (when faved
+               (format "(%s) " (propertize "F" 'face 'success)))
+             (mastodon-tl--byline-author toot)
+             (mastodon-notifications--byline-concat toot message)
+             (propertize "\n  ------------" 'face 'default))
+     'toot-id id
+     'toot-json toot)))
+
+(defun mastodon-notifications--mention (note)
+  "Format for a `mention' NOTE."
+  (let ((toot (mastodon-tl--field 'status note)))
+    (insert
+     (concat (mastodon-tl--content toot)
```

Comment:
`concat` not needed (we make that mistake in various places).

`insert` takes any amount of args and inserts them in order, AFAIK.